src/deserialize.zig

```diff
@@ -114,10 +115,16 @@ pub fn deserialize(comptime T: type, serialized: []const u8, out: *T) !usize {
 
                 const r = sizeAndDataOffset(serialized);
                 var end = r.offset + r.size;
+
+                // since it's not possible to say if the slice is "undefined", it has
+                // to be allocated regardless.
+                out.* = &[_]ptr.child{};
+
                 var offset = r.offset;
                 var i: usize = 0;
                 while (offset < end) : (i += 1) {
-                    offset += try deserialize(ptr.child, serialized[offset..], &out.*[i]);
+                    out.* = try allocator.realloc(out.*, (i + 1));
```

Comment:
not super happy to reallocate every time, but the alternative is to go over the payload and rewind, which I could do but not in this PR as I think it will turn out to be quite complicated, depending on the underlying structure.

```suggestion
                    // TODO create a helper that counts the number of elements
                    out.* = try allocator.realloc(out.*, (i + 1));
```

Reply:
Uhm, yeah, that would be the ideal thing.

Some middle point is to take a similar policy as Go slices or ArrayList growth. Every time you run out of space, you double (or 1.25x) the slice. So you kind of amortize the amount of allocs. 

The tradeoff is that the resulting slice might be slightly bigger than needed -- but that's usually accepted in most languages (Go and Zig "growable" slices).

In any case, what you mention reg continue reading to count the num of elements sounds like the optimal take on the problem for sure. (with some extra CPU work)